src/CovidSim.cpp

```diff
@@ -896,10 +894,8 @@ void ReadParams(char* ParamFile, char* PreParamFile)
 				s = ((double)i) / 1024;
 				while (P.JourneyDurationDistrib[j] < s)j++;
 				P.InvJourneyDurationDistrib[i] = j;
-			}
-			for (i = j = 0; i <= 1024; i++)
-			{
-				s = ((double)i) / 1024;
+
+				j = 0;
 				while (P.LocalJourneyDurationDistrib[j] < s)j++;
 				P.InvLocalJourneyDurationDistrib[i] = j;
 			}
```

Comment:
This will be much slower, as you're resetting `j` in each iteration. `j` is stepping through, I think, cumulative distributions (see their calculation a bit earlier) so you can leave `j` where it is and continue searching onwards.

If you *really* want to merge the loops, I think you need...
for (i = j1 = j2 = 0; i <= 1024; i++) 
{
    s = ((double)i) / 1024;
    while (P.JourneyDurationDistrib[j1] < s) j1++;
    P.InvJourneyDurationDistrib[i] = j1;
    while (P.LocalJourneyDurationDistrib[j2] < s) j2++;
    P.InvLocalJourneyDurationDistrib[i] = j2;
}

Reply:
Can you benchmark that somehow? On the other side, the code is looped 1024 instead of 2048 times, and on top of that, i isn't calculated twice. I think the benefits of that are pretty clear.

Reply:
As @weshinsley pointed out, `P.JourneyDurationDistrib` and `P.LocalJoureyDurationDistrib` are both cumulative distribution functions (i.e. the numbers will be ordered from smallest to largest). If you reset `j = 0` you are repeating unnecessary comparisons up until the previous `j` value search since `P.JourneyDurationDistrib` at every value before `j` will be less than `(i + 1) / 1024`. The outer loop may have double the iterations, but the inner loop is what will be calculated the most if `j` were reset every time.

I think that the merged loop is the correct refactor here. It halves the number of outer loop iterations while keeping the number of inner loop iterations the same. The only cost is an additional integer variable. However to be completely frank, this is a premature optimization. I've been profiling the code this week (see #388) and my analysis shows that the entire `ReadParams()` function takes less than 0.1% of the run-time.

Reply:
Yes, I agree, this section isn't one to overthink about performance - and also I don't think this PR was intended to be about performance - it's really a bite-sized code clean-up, where the review highlighted ways it behaved differently (on both behaviour and performance) to the original.

But all looks good to me now.

Reply:
> Is that solution slower than the current one (before my change)? I would proceed with using that.

As a serious answer to this question: Who knows?

These loop merging PRs concern me because they change memory access patterns - and therefore how the CPU will feed the "beast" that is the cache and memory hierarchy.

The original loops were simple - two arrays being accessed sequentially in each loop.  I would expect a server-class CPU to be able to handle this and be able to track the accesses patterns to request memory far enough ahead of time that we won't get stalls in the memory system.  However, merging these loops means we're interleaving more accesses.  I would expect some CPUs to start to struggle with this and the code to actually be slower.

CovidSim's biggest performance bottle-neck is memory accesses - real runs use far more memory than available in the caches.  So giving the CPU all the help it can to predict memory usage is likely to be best way to improve performance.  This involves things like thinking about structure layout, data locality, and access patterns.

Having said that - in this case we're doing initialisation in cold code so performance is not as much an issue as readibability.

Reply:
@matt-gretton-dann If this change is problematic, I can propose the one at lines 675 only.

Reply:
No I'm fine with this as it is - I'm just trying to raise awareness that even simple changes can/will have significant knock on effects around the performance of this code.